sycl/doc/DeviceGlobal.md

```diff
@@ -0,0 +1,602 @@
+
+#### Initializing the device global variables in device code
+
+When a DPC++ application submits a kernel, the runtime constructs a
+`pi_program` containing this kernel that is compiled for the target device, if
+such a `pi_program` does not yet exist.  If the kernel resides in a device code
+module that calls into a shared library, the runtime identifies a set of device
+code modules that need to be online-linked together in order to construct the
+`pi_program`.
+
+After creating a `pi_program` and before invoking any kernel it contains, the
+runtime does the following:
+
+* Scan the entries of the "SYCL/device globals" property sets of each device
+  code module that contributes to the the `pi_program` to get information about
+  each device global variable that is used by the `pi_program`.  This
+  information is added to device global database.
+
+* For each device global variable that is not decorated with the
+  `device_image_scope` property:
+
+    - Check if a USM buffer has already been created for the variable on this
+      target device.  If not, the runtime allocates the buffer from USM device
+      memory using the size from the database and zero-initializes the content
+      of the buffer.  The pointer to this buffer is saved in the database for
+      future reuse.
+
+    - Regardless of whether the USM buffer has already been created for the
+      variable, the runtime initializes the `usmptr` member in the *device
+      instance* of the variable by using a backend-specific function which
+      copies data from the host to a device variable.  It is a simple matter to
+      use this function to overwrite the `usmptr` member with the address of
+      the USM buffer.  The details of this device-specific function are
+      described below.
+
+Note that the runtime does not need to initialize the `val` member variable of
+device global variables that are decorated with `device_image_scope` because
+the constructor already guarantees that this member variable is zero
+initialized.
+
+#### Implementing the `copy` and `memcpy` functions in `queue` and `handler`
+
+Each of these functions accepts a (host) pointer to a device global variable as
+one of its parameters, and the runtime uses this pointer to find the associated
+information for this variable in the device global database.  In the common
+case where a kernel using this device global has already been submitted to the
+target device, the database will contain all of the information for this device
+global variable.
+
+However, in the case when the application has not previously submitted a kernel
+that uses this device global, the database will contain only the address of the
+variable and its unique string.  In this case, the runtime must scan all
+"SYCL/device globals" property sets in the application searching for an entry
+with that same unique string.  The runtime can then add the remaining
+information about the device global variable to the database.
```

Comment:
Probably, we need to clarify what happens when there are several(e.g. compiled with different values for spec constants) already created program which use a `device_global`. And when an additional program is created after the copy call.

Reply:
> Probably, we need to clarify what happens when there are several(e.g. compiled with different values for spec constants) already created program which use a device_global.

This is a good point.  I think the extension API spec should be updated to describe two error cases for the `copy` and `memcpy` functions:

* If either of these functions attempts to read or write beyond the end of the specified device global variable, the function throws an exception with the `errc::invalid` error code.

* If either of these functions is called for a device global variable that has the `device_image_scope` property and the variable exists in more than one *device image*, then the function throws an exception with the `errc::invalid` error code.

The implementation should distinguish the two cases with different `what` strings in the exception.

Note that the second error condition can occur if the application executes a kernel with different values of a specialization constant (in the case when the device does not have native support for specialization constants).  It can also occur if the application creates two or more kernel bundles via the `kernel_bundle` API.

> And when an additional program is created after the copy call.

This is also a good point.  I realize now that the `piextCopy{To,From}DeviceVariable` functions require an additional `pi_program` parameter.  This implies that the `copy` / `memcpy` functions will need to create a `pi_program` (if one does not already exist) only in the case when the device global variable has the `device_image_scope` property.

Reply:
I believe this is addressed now.  I clarified the spec in #4972 and commit ba65c99a9ce27f32254acbc857b81eaeae1aab4d in this PR updates the design for the copy functions.  Thanks for pointing out this problem!